packages/Sandblocks-Core/SBForceMoveDecorator.class.st

```diff
@@ -84,6 +100,27 @@ SBForceMoveDecorator >> blockAtPoint: aPoint [
 	^ best
 ]
 
+{ #category : #'as yet unclassified' }
+SBForceMoveDecorator >> changedDirection [
+	
+	^ velocity isZero or:
+		[self previousVelocity isZero] or:
+		[(velocity angleWith: self previousVelocity) between: Float pi * 0.5 and: Float pi * 1.5]
+]
+
+{ #category : #'as yet unclassified' }
+SBForceMoveDecorator >> createImplicitForces [
+	
+	self morph owner submorphsReverseDo: [:otherMorph |
+		(otherMorph isSandblock and: [
+			otherMorph preventOcclusion and: [
+				otherMorph ~= self morph and: [
+					otherMorph isDragging not]]]) ifTrue: [ | attraction |
+						attraction := self attractionTo: otherMorph.
+						attraction > 0 ifTrue: [
+							self morph sandblockEditor openMorphBack: (SBForceConnector connect: self morph to: otherMorph forceScale: attraction)]]].
```

Comment:
ah interesting that this worked reliably, technically you're modifying a collection while you're iterating over it. The safe way to go would be to either queue adding the connectors or creating a shallow copy of the submorph collection, e.g. `self morph owner submorphs reversed do:`. You could also consider switching to `self morph sandblockEditor childSandblocks "already creates copy" revserseDo:`, which is also semantically a little cleaner.